### PR TITLE
API: Introduce ``np.astype`` [Array API]

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @rgommers @ngoldbaum,

This PR adds `np.astype` for Array API compatibility (https://data-apis.org/array-api/latest/API_specification/generated/array_api.astype.html).

It adheres to Array API and uses default values for `ndarray.astype` arguments outside of the standard. 

Tracking issue: https://github.com/numpy/numpy/issues/25076

### CODE REVIEW

File: numpy/_core/numeric.py [hunk 1]
```diff
@@ -2508,6 +2507,60 @@ def array_equiv(a1, a2):
...
+
+    Non-copy case:
+
+    >>> arr = np.array([1, 2, 3])
+    >>> arr_noncpy = np.astype(arr, arr.dtype, copy=False)
+    >>> np.shares_memory(arr, arr_noncpy)
+    True
+
+    """
+    x = asarray(x)
```
Comment (seberg) [inline numpy/_core/numeric.py:2560]: This isn't really right:
* You would have to use `copy` in the `asarray` call.
* The `copy=copy` in `astype()` doesn't have the same semantics.

I am actually not sure if `astype` should *coerce* from to an array at all?  If this does a forced cast, then `astype` *becomes* `asarray()`.

Comment (mtsokol) [reply numpy/_core/numeric.py:2560]: I also wonder if any coercion should be allowed here and I would agree if it shouldn't.

Comment (rgommers) [reply numpy/_core/numeric.py:2560]: I think I agree about no coercion here. Instead, it seems to me like this should verify that the input is already an array.

That would leave the question of how exactly to do that. The dispatcher for `__array_function__` was already added, so I think it should work to check with `isinstance(x, np.ndarray)` (allows subclasses, but not duck arrays - those should be using `__array_function__`).

And, if we do that, what's the right exception type? I'd think `TypeError`?

Comment (mtsokol) [reply numpy/_core/numeric.py:2560]: I second `TypeError`. Its documentation says:
```
Raised when an operation or function is applied to an object of inappropriate type.

File: numpy/_core/numeric.pyi [hunk 2]
```diff
@@ -658,3 +658,9 @@ def isclose(
 def array_equal(a1: ArrayLike, a2: ArrayLike, equal_nan: bool = ...) -> bool: ...
 
 def array_equiv(a1: ArrayLike, a2: ArrayLike) -> bool: ...
+
+def astype(
+    x: ArrayLike,
+    dtype: DTypeLike,
+    copy: bool = ...,
+) -> NDArray[Any]: ...
```
Comment (BvB93) [inline numpy/_core/numeric.pyi:666]: `astype` could really use an overload here for all `generic`-parametrizable dtype-likes, as we can quite easily statically express the dtype of the returned arrays in those cases.
```suggestion
@overload
def astype(
    x: ArrayLike,
    dtype: _DTypeLike[_SCT],
    copy: bool = ...,
) -> NDArray[_SCT]: ...
@overload
def astype(
    x: ArrayLike,
    dtype: DTypeLike,
    copy: bool = ...,
) -> NDArray[Any]: ...
```

Comment (mtsokol) [reply numpy/_core/numeric.pyi:673]: I committed it - thanks!

File: numpy/_core/numeric.py [hunk 3]
```diff
@@ -2517,6 +2516,63 @@ def array_equiv(a1, a2):
...
+def astype(x, dtype, /, *, copy = True):
+    """
+    Copies an array to a specified data type.
+
+    This function is an Array API compatible alternative to
+    `numpy.ndarray.astype`.
+
+    Parameters
+    ----------
+    x : array_like
```
Comment (ngoldbaum) [inline numpy/_core/numeric.py:2533]:
```suggestion
    x : `ndarray`
```

Maybe also a note that array-likes are explicitly not supported here?

Comment (mtsokol) [reply numpy/_core/numeric.py]: Right, I updated it to `x : ndarray` to follow existing docs in other functions.
I also added the note.

**Comment by ngoldbaum** [comment]
Thanks @mtsokol!

**Pull request merged**